internal/sharedcomponent/sharedcomponent.go

```diff
@@ -12,23 +13,25 @@ import (
 	"go.opentelemetry.io/collector/component"
 )
 
-// SharedComponents a map that keeps reference of all created instances for a given configuration,
-// and ensures that the shared state is started and stopped only once.
-type SharedComponents[K comparable, V component.Component] struct {
-	comps map[K]*SharedComponent[V]
+// Map keeps reference of all created instances for a given shared key such as a component configuration.
+type Map[K comparable, V component.Component] interface {
+	// LoadOrStore returns the already created instance if exists, otherwise creates a new instance
+	// and adds it to the map of references.
+	LoadOrStore(key K, create func() (V, error), telemetrySettings *component.TelemetrySettings) (*Component[V], error)
 }
 
-// NewSharedComponents returns a new empty SharedComponents.
-func NewSharedComponents[K comparable, V component.Component]() *SharedComponents[K, V] {
-	return &SharedComponents[K, V]{
-		comps: make(map[K]*SharedComponent[V]),
+func NewMap[K comparable, V component.Component]() Map[K, V] {
+	return &mapImpl[K, V]{
+		components: map[K]*Component[V]{},
 	}
 }
 
-// GetOrAdd returns the already created instance if exists, otherwise creates a new instance
-// and adds it to the map of references.
-func (scs *SharedComponents[K, V]) GetOrAdd(key K, create func() (V, error), telemetrySettings *component.TelemetrySettings) (*SharedComponent[V], error) {
-	if c, ok := scs.comps[key]; ok {
+type mapImpl[K comparable, V component.Component] struct {
```

Comment:
Separate PR we may consider to do this map thread safe.

Reply:
Sure thing!

Reply:
#9170 is ready for after this is merged.